### ISSUE TITLE
Using batch-wise maximum sequence lengths for padding

### ISSUE THREAD

**Issue by sayakpaul** [keras-team/keras#15810]
If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.

TensorFlow version (you are using): 2.7.0
Are you willing to contribute it (Yes/No) : No

**Describe the feature and the current behavior/state**.

In NLP, padding and truncation are common tasks in order to have sequences inside a batch to be of uniform lengths. A common practice is to pad all the shorter sequences in a dataset to the maximum allowable sequence length by a model (for BERT, it's usually 512) and truncate the sequences that are shorter than that maximum length. 

But this leads to a lot of compute waste since not all batches will have a maximum sequence length of 512, for example. The batch-wise maximum sequence length may very well be shorter than 512. So, it's a better practice to pad with respect to `min(batchwise_max_seq_len, max_seq_len_for_the_model)`. [This article](https://mccormickml.com/2020/07/29/smart-batching-tutorial/) details this practice. The figure below (taken from the article) beautifully depicts this approach:

![image](https://user-images.githubusercontent.com/22957388/146672004-f52832bf-5443-47d5-80e2-3898d74885d0.png)

My question is **_how do we do this in Keras particularly for models where we want to use simpler techniques in the preprocessing pipeline?_** The model could be a shallow MLP and let's say we are presented with a text classification problem. For tokenization, we have TF+IDF (doable via `keras.layers.preprocessing`) but I haven't been able to figure a way out to do the padding and truncation using the compute-efficient approach I described above. 

[`bucket_by_sequence_length()`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#bucket_by_sequence_length) seems like a promising approach but I am not sure how it would fit in with the preprocessing layers offered off-the-shelf by Keras. I did start some work but couldn't make much progress. Here's my starter [Jupyter Notebook](https://gist.github.com/sayakpaul/4d46ff36bd0975efc99e8e884c050138). 

@mattdangerw any ideas? 

**Will this change the current api? How?**

It probably won't change the API if we can figure a way out with the existing methods. 

**Who will benefit from this feature?**

NLP practitioners. 

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): No. I can contribute an example for keras.io. 
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

P.S.: At my organization (Carted), we have figured a way out to prepare the sequences using the compute-efficient padding approach but it's very specific for a type of BERT model we are using. We are currently working on a blog post to share our approach with the community.

**Comment by chjort** [comment]
This can be done quite elegantly if you cast your batch to a `tf.RaggedTensor`.

```python
batch = [
    ["a", "b", "c"],
    ["a", "b", "c", "d", "e"],
    ["a", "b"]
]
ragged_batch = tf.ragged.constant(batch)

pad_value = tf.constant("[PAD]")
shape = ragged_batch.bounding_shape()
ragged_batch = ragged_batch.to_tensor(default_value=pad_value, shape=shape)
```

This gives
```
<tf.Tensor: shape=(3, 5), dtype=string, numpy=
array([[b'a', b'b', b'c', b'[PAD]', b'[PAD]'],
       [b'a', b'b', b'c', b'd', b'e'],
       [b'a', b'b', b'[PAD]', b'[PAD]', b'[PAD]']], dtype=object)>
```

**Comment by sayakpaul** [comment]
@chjort this is not padding to the maximum sequence length present in a batch. All the batches (in your example) still have uniform sequence lengths. Consider the following figure to understand what I am trying to mean:

<img width="596" alt="variable_length_batch" src="https://user-images.githubusercontent.com/22957388/147175705-3613c83e-dea9-46ac-8af1-49e65a1a11a1.png">
 
On the left-hand side, we have a bunch of sequences. After batching, we can notice that a batch of sequences has been padded (when necessary) to the maximum sequence length particular to that batch.

**Comment by chjort** [comment]
Well this can easily be achieved using my first suggestion with the `tf.RaggedTensor` and a `tf.data.Dataset`:

```python
data = [
    [1, 2, 3, 4, 5],
    [1, 2, 3],
    [1, 2, 3, 4, 5, 6],
    [1, 2, 3, 4, 5],
    [1, 2],
    [1, 2, 3, 4],
    [1, 2, 3, 4, 5, 6, 7],
    [1, 2, 3],
    [1],
    [1, 2, 3],
    [1, 2, 3, 4, 5, 6],
    [1, 2, 3, 4, 5],
    [1, 2],
    [1, 2, 3, 4, 5, 6, 7],
    [1, 2, 3, 4, 5],
]
data_ragged = tf.ragged.constant(data)
dataset_ragged = tf.data.Dataset.from_tensor_slices(data_ragged)
dataset_ragged = dataset_ragged.batch(5)


def batch_pad(batch: tf.RaggedTensor):
    pad_value = 0
    return batch.to_tensor(default_value=pad_value)


dataset_padded = dataset_ragged.map(batch_pad)

for batch in dataset_padded:
    print(batch.shape)
```
yielding the following batch shapes:
```
(5, 6)
(5, 7)
(5, 7)
```
exactly like in your above figure.

**Comment by haifeng-jin** [comment]
summary: what is the most elegant way to pad sequences with to max len of each batch instead of the global max len.

**Comment by sayakpaul** [comment]
@haifeng-jin, 

@Nilabhra and I have worked on a series of blog posts that discuss this issue in detail and propose different solutions. We haven't yet published those blog posts (pending approval from the company). As soon as they are published we will let you all know here.

**Comment by sayakpaul** [comment]
@mattdangerw @haifeng-jin 

We have open-sourced our repository that dives deep into handling variable-sized sequences lengths in TensorFlow and Keras: https://github.com/carted/handling-variable-length-text-tf

Cc: @Nilabhra
